test/box/net.box_closing_without_lost_gh-6338.test.lua

```diff
@@ -0,0 +1,164 @@
+    local conn = net_box.connect(addr)
+    local stream = conn:new_stream()
+    local space = stream.space.test
+    stream:begin()
+    for i = 1, count do space:replace({i}, {is_async = true}) end
+    stream:commit({is_async = true})
+    conn:close()
+    test_run:wait_cond(function ()
+        return get_current_connection_count() == 0
+    end)
+end;
+function create_remote_space(engine)
+    test_run:eval('test',
+        string.format(
+            "s = box.schema.space.create('test', { engine = '%s' })",
+            engine
+        )
+    )
+    test_run:eval('test', string.format("s:create_index('primary')"))
+end;
+test_run:cmd("setopt delimiter ''");
+
+-- Some simple checks for new object - stream
+test_run:cmd("start server test with args='10, true'")
+test_run:switch("test")
+test_run:cmd("setopt delimiter ';'")
+function check_requests_result_and_drop_space(space, count)
+    local errinj = box.error.injection
+    assert(errinj.get('ERRINJ_IPROTO_STREAM_COUNT') == 0)
+    assert(errinj.get('ERRINJ_IPROTO_STREAM_MSG_COUNT') == 0)
+    local rc = space:select()
+    space:drop()
+    return (#rc == count)
+end;
+test_run:cmd("setopt delimiter ''");
+test_run:switch("default")
+
+server_addr = test_run:cmd("eval test 'return box.cfg.listen'")[1]
+net_msg_max = test_run:cmd("eval test 'return box.cfg.net_msg_max'")[1]
+create_remote_space("memtx")
+connection_send_async_requests(net_msg_max - 1, server_addr, true)
+test_run:switch("test")
+assert(check_requests_result_and_drop_space(s, box.cfg.net_msg_max - 1))
+test_run:switch("default")
+create_remote_space("vinyl")
+connection_send_async_requests(net_msg_max - 1, server_addr, true)
+test_run:switch("test")
+assert(check_requests_result_and_drop_space(s, box.cfg.net_msg_max - 1))
+test_run:switch("default")
+create_remote_space("memtx")
+stream_send_async_requests(net_msg_max - 3, server_addr)
+test_run:switch("test")
+assert(check_requests_result_and_drop_space(s, box.cfg.net_msg_max - 3))
+test_run:switch("default")
+create_remote_space("vinyl")
+stream_send_async_requests(net_msg_max - 3, server_addr)
+test_run:switch("test")
+assert(check_requests_result_and_drop_space(s, box.cfg.net_msg_max - 3))
+test_run:switch("default")
+create_remote_space("memtx")
+
+big_msg_count = 100 * net_msg_max
+-- A special test to check that there is no hang when the connection is
+-- closed from the server side, when not all requests have been sent yet.
+-- (see send_cond:broadcast() usage from net_box.lua)
+cond = fiber.cond()
+channel = fiber.channel(1)
+test_run:cmd("setopt delimiter ';'")
+_ = fiber.create(function()
+    cond:wait()
+    test_run:cmd("stop server test")
+    channel:put(true)
+end);
+test_run:cmd("setopt delimiter ''");
+-- We are sending a very large number of requests here to ensure that
+-- they will not have time to send by the time the connection is closed.
+connection_send_async_requests(big_msg_count, server_addr, false, cond)
```

Comment:
The test is unexpectedly huge. You don't need to send many messages or use streams here. You can block the worker fiber with ERRINJ_NETBOX_IO_DELAY and send just a few messages. Please try to simplify it.

Reply:
Fixed in all places expect one special test case, where we check that there is no hang when the connection is
closed from the server side. In this test case we need to send large count of messages, but it doesn't take many time

Reply:
Also i would like to note, that it has one significant disadvantage: errinj not work in release CI!